### PR TITLE
xenserver_guest - fixed validations

### PR DESCRIPTION
**Pull request by russoz**
##### SUMMARY
Fixed sanity checks for module.

##### ISSUE TYPE
- Docs Pull Request

##### COMPONENT NAME
plugins/modules/cloud/xenserver/xenserver_guest.py

### CODE REVIEW

File: plugins/modules/cloud/xenserver/xenserver_guest.py
Comment (bvitnik) [inline plugins/modules/cloud/xenserver/xenserver_guest.py:1876]: Not required. `iso_name` can be set without `type` in which case `type` is implicitly set to `iso`. If nothing is set, `type` is assumed to be `none`. This seems to make `type` useless but `type` is kept as is for now because there is an unimplemented type `host`. Once support for `host` type is implemented, `type` will have to be explicitly set.

File: plugins/modules/cloud/xenserver/xenserver_guest.py
Comment (bvitnik) [inline plugins/modules/cloud/xenserver/xenserver_guest.py:1893]: None are required. `name` is not required for existing network interfaces. For example you can change a MAC address only for some interface. `type` is not required because in some cases value is assumed depending on other parameters that are specified.

**Review by bvitnik — CHANGES_REQUESTED** [review]

**Comment by bvitnik** [comment]
Additional comment. I'm not sure why non `xenserver_guest` related lines in `tests/sanity/ignore-*.txt` files are removed. That will surely break sanity checks for other modules.

File: plugins/modules/cloud/xenserver/xenserver_guest.py
Comment (russoz) [reply plugins/modules/cloud/xenserver/xenserver_guest.py:144]: yup, coming up in the next push

File: plugins/modules/cloud/xenserver/xenserver_guest.py
Comment (russoz) [reply plugins/modules/cloud/xenserver/xenserver_guest.py:1837]: There was a `required: true` in the documentation - which I removed, but left the question here just in case.

File: plugins/modules/cloud/xenserver/xenserver_guest.py
Comment (russoz) [reply plugins/modules/cloud/xenserver/xenserver_guest.py:184]: Done.

File: plugins/modules/cloud/xenserver/xenserver_guest.py
Comment (bvitnik) [reply plugins/modules/cloud/xenserver/xenserver_guest.py:1837]: That's OK. `required: true` in the documentation was an error.

File: plugins/modules/cloud/xenserver/xenserver_guest.py
Comment (russoz) [reply plugins/modules/cloud/xenserver/xenserver_guest.py:1861]: done

File: plugins/modules/cloud/xenserver/xenserver_guest.py
Comment (russoz) [reply plugins/modules/cloud/xenserver/xenserver_guest.py:1876]: Shouldn't we have a `default="iso"` until then?

File: plugins/modules/cloud/xenserver/xenserver_guest.py
Comment (russoz) [reply plugins/modules/cloud/xenserver/xenserver_guest.py:1893]: done

File: plugins/modules/cloud/xenserver/xenserver_guest.py
Comment (bvitnik) [reply plugins/modules/cloud/xenserver/xenserver_guest.py:1876]: No. If `type` has value `iso`, than `iso_name` must be specified. There is no way to tell Ansible "I want an empty CDROM", or better said, you would always have to specify `type: none`.

**Comment by russoz** [comment]
> Additional comment. I'm not sure why non `xenserver_guest` related lines in `tests/sanity/ignore-*.txt` files are removed. That will surely break sanity checks for other modules.

I had a semi-automated script to help filter out the modules from these files - and it did work well before, but I  must have obviously made some silly mistake and of course I didn't double check this time. Gotta write tests for all these things. Anyways, fixed that, let me just wait for the CI results.

Thanks for the prompt review!

File: plugins/modules/cloud/xenserver/xenserver_guest.py
Comment (russoz) [reply plugins/modules/cloud/xenserver/xenserver_guest.py]: Anyway, that should not be in this PR. :-)

**Comment by bvitnik** [comment]
No problem. Let's wait for CI results.

I'll maybe ask for additionally changes in documentation as I have a PR of my own still in preparation which dealt with sanity checks but I never got time to commit. So it would be beneficial to for both PRs to be in line. I'll get back to you once I do a diff.

**Comment by russoz** [comment]
Looking good. Do you want to put the additional changes in here (== should we wait to merge this PR) or would you rather raise another one? Works the same for me. Cheers,

**Comment by bvitnik** [comment]
Better additional changes in this one. Less work for me :). I'll get back to you tomorrow regarding this.

File: plugins/modules/cloud/xenserver/xenserver_guest.py
Comment (felixfontein) [inline plugins/modules/cloud/xenserver/xenserver_guest.py:199]:
```suggestion
          description: Static IPv4 netmask required for I(ip) if prefix is not specified.
```

File: plugins/modules/cloud/xenserver/xenserver_guest.py
Comment (felixfontein) [inline plugins/modules/cloud/xenserver/xenserver_guest.py:209]:
```suggestion
          description: 'Static IPv6 address (implies I(type6): C(static)) with prefix in format C(<IPv6 address>/<prefix>).'
```

File: plugins/modules/cloud/xenserver/xenserver_guest.py
Comment (bvitnik) [inline plugins/modules/cloud/xenserver/xenserver_guest.py:63]: I believe that by new guidelines this line:

`from __future__ import absolute_import, division, print_function`

should be:

`from __future__ import (absolute_import, division, print_function)`

File: plugins/modules/cloud/xenserver/xenserver_guest.py
Comment (bvitnik) [inline plugins/modules/cloud/xenserver/xenserver_guest.py:109]: This is just a suggestion, not a request. In my code I use this style:

```
suboptions:
  num_cpus:
    description:
    - Number of CPUs.
```

Same for other one line descriptions.

File: plugins/modules/cloud/xenserver/xenserver_guest.py
Comment (bvitnik) [inline plugins/modules/cloud/xenserver/xenserver_guest.py:123]: This should read:

```- VM needs to be shut down to reconfigure **disk size**.```

If you say `this parameter` it is misleading. It could mean that reconfiguring anything in `disks:` requires VM to be shut down which is not true. Adding new disks and renaming existing ones can be done online. Resizing is the problem.

File: plugins/modules/cloud/xenserver/xenserver_guest.py
Comment (bvitnik) [inline plugins/modules/cloud/xenserver/xenserver_guest.py:177]: Add this one line:

```- Name is required for new NICs. Other parameters are optional in all cases.```

File: plugins/modules/cloud/xenserver/xenserver_guest.py
Comment (bvitnik) [inline plugins/modules/cloud/xenserver/xenserver_guest.py:228]: Change description to:

`VM param name.`

File: plugins/modules/cloud/xenserver/xenserver_guest.py
Comment (bvitnik) [inline plugins/modules/cloud/xenserver/xenserver_guest.py:232]: Change description to:

`VM param value.`

**Review by bvitnik — CHANGES_REQUESTED** [review]

**Comment by bvitnik** [comment]
@russoz ... and thanks for all the effort!

**Comment by russoz** [comment]
@bvitnik @felixfontein  I think I covered the pending issues. Please review.

File: plugins/modules/cloud/xenserver/xenserver_guest.py
Comment (bvitnik) [inline plugins/modules/cloud/xenserver/xenserver_guest.py:125]: Change to:

`New disks are required to have either a I(size) or one of I(size_[tb,gb,mb,kb,b]) parameters specified.`

It makes things more clear as size iz not required for existing disks.

File: plugins/modules/cloud/xenserver/xenserver_guest.py
Comment (bvitnik) [inline plugins/modules/cloud/xenserver/xenserver_guest.py:133]: Add line:

`If no unit is specified, size is assumed to be in bytes.`

File: plugins/modules/cloud/xenserver/xenserver_guest.py
Comment (bvitnik) [inline plugins/modules/cloud/xenserver/xenserver_guest.py:186]: Fix indentation. There are two extra spaces compared to the rest of the documentation.

**Review by bvitnik — CHANGES_REQUESTED** [review]

**Comment by bvitnik** [comment]
That should be all from me. We can merge this as soon as these changes are commited.

**Review by felixfontein — APPROVED** [review]

File: plugins/modules/cloud/xenserver/xenserver_guest.py
Comment (felixfontein) [reply plugins/modules/cloud/xenserver/xenserver_guest.py]: @bvitnik which guidelines do you mean?

**Comment by felixfontein** [comment]
@bvitnik for auto-merge you would have to give an Approve review first, because right now you have a Change Requested review still active :) But since I'll just merge manually, that doesn't matter now.

**Comment by felixfontein** [comment]
@russoz thanks for fixing this!
@bvitnik thanks for reviewing!

**Comment by felixfontein** [comment]
The automatic backport to stable-1 failed, if someone wants these improvements also there, you need to do a manual backport.

**Comment by russoz** [comment]
@bvitnik given there's been no code change (docs only), I do not see the need to work on backporting it.

File: plugins/modules/cloud/xenserver/xenserver_guest.py
Comment (bvitnik) [reply plugins/modules/cloud/xenserver/xenserver_guest.py]: All the examples in Ansible Developer Guide use this form. I can also vaguely recall that there was a massive commit to change this line in some collection of modules. It prompted me to also change the line. I'm really not sure any more :).

**Comment by bvitnik** [comment]
Unfortunately I haven't closely followed Ansible (and this collection) development for quite a while now. So I'm not sure how things work any more and what stable-1 and 2 branches are used for. I'll let you decide :).

File: plugins/modules/cloud/xenserver/xenserver_guest.py
Comment (felixfontein) [reply plugins/modules/cloud/xenserver/xenserver_guest.py]: Ah ok. The sanity tests accept both versions, so I guess both are ok. I just looked at ansible/ansible, there every file in lib/ansible uses `from __future__ import (absolute_import, division, print_function)`, **except** all modules, some module_utils, plugins/filter/urls.py and plugins/doc_fragments/vars_plugin_staging.py: these use `from __future__ import absolute_import, division, print_function`. So... no idea :)

**Comment by felixfontein** [comment]
@bvitnik while 3.0.0 will be released in ~a week, we still release bugfix releases 1.3.x and 2.5.y from time to time. I have no idea whether people actually use them... In any case, what we usually do is always backport import fixes (like security fixes) if possible, and for other fixes it depends on whether the bot manages to auto-backport them or not. If the bot doesn't do the work, it really depends on whether there's interest (or whether we care enough)... Now that we're almost at major version 3, I would still try to backport to stable-2, but I wouldn't invest work for stable-1 if the fix is not important and nobody asks for it. So in this case, I would leave it as-is :)

**Pull request merged**
